rust/sbp/src/de.rs

```diff
@@ -196,36 +189,80 @@ impl Decoder for SbpDecoder {
             }
         };
         src.advance(start);
-        match parse_frame(src) {
-            Some(Ok(frame)) => Ok(Some(Sbp::from_frame(frame)?)),
-            Some(Err(err)) => Err(err.into()),
-            None => {
-                src.reserve(MAX_FRAME_LEN);
-                Ok(None)
-            }
+        if src.len() < HEADER_LEN {
+            src.reserve(MAX_FRAME_LEN);
+            return Ok(None);
         }
+        let payload_len = src[5] as usize;
+        let at = HEADER_LEN + payload_len + CRC_LEN;
+        if src.len() < at {
+            src.reserve(MAX_FRAME_LEN);
+            return Ok(None);
+        }
+        let frame = SbpFrame(src.split_to(at));
+        Ok(Some(frame))
     }
 
-    fn decode_eof(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
-        let res = match self.decode(buf) {
-            Ok(Some(frame)) => Ok(Some(frame)),
-            _ => Ok(None),
-        };
-        buf.clear();
-        res
+    fn decode_eof(&mut self, _src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        todo!()
+    }
+}
+
+/// The wire representation of an SBP message.
+#[derive(Debug)]
+pub struct SbpFrame<B>(B);
+
+impl<B: bytes::Buf> SbpFrame<B> {
```

Comment:
Might be nice to add a `pub fn as_bytes(&self) -> &[u8] { self.0.chunk() }` - would clean up these other helpers and might be nice for an end user

Reply:
Hmm or maybe `AsRef<[u8]>`